src/aws-cpp-sdk-core/include/aws/core/utils/stream/StreamBufProtectedWriter.h

```diff
@@ -0,0 +1,101 @@
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#pragma once
+
+#include <aws/core/Core_EXPORTS.h>
+#include <aws/core/utils/Array.h>
+#include <streambuf>
+#include <functional>
+
+namespace Aws
+{
+    namespace Utils
+    {
+        namespace Stream
+        {
+            /**
+             * This is a wrapper to perform a hack to write directly to the put area of the underlying streambuf
+             */
+            class AWS_CORE_API StreamBufProtectedWriter : public std::streambuf
+            {
+            public:
+                StreamBufProtectedWriter() = delete;
+
+                using WriterFunc = std::function<bool(char* dst, uint64_t dstSz, uint64_t& read)>;
+
+                static size_t ForceOverflow(StreamBufProtectedWriter* pBuffer, const WriterFunc& writerFunc)
+                {
+                    char dstChar;
+                    size_t read = 0;
+                    if (writerFunc(&dstChar, 1, read) && read > 0)
+                    {
+                        pBuffer->overflow(dstChar);
+                        return 1;
+                    }
+                    return 0;
+                }
+
+                static size_t WriteToBuffer(std::streambuf* pBuffer, const WriterFunc& writerFunc)
+                {
+                    StreamBufProtectedWriter* pBufferCasted = static_cast<StreamBufProtectedWriter*>(pBuffer);
+                    auto dstBegin = pBufferCasted->pptr();
+                    auto dstSz = pBufferCasted->epptr() - dstBegin;
+                    size_t totalRead = 0;
+                    uint64_t read = 0;
+                    if (dstSz == 0)
+                    {
+                        // prime the initial buffer
+                        char tmpBuf[1024];
+                        uint64_t tmpBufSz = sizeof(tmpBuf);
+                        if (writerFunc(tmpBuf, tmpBufSz, read) && read > 0)
+                        {
+                            pBufferCasted->xsputn(tmpBuf, read);
+                            totalRead += read;
+                            if (pBufferCasted->pptr() >= pBufferCasted->epptr())
+                            {
+                                if(!ForceOverflow(pBufferCasted, writerFunc))
+                                {
+                                    return totalRead;
+                                } else {
+                                    totalRead++;
+                                }
+                            }
+                        } else {
+                            return 0;
+                        }
+
+                        dstBegin = pBufferCasted->pptr();
+                        assert(pBufferCasted->epptr() > dstBegin);
+                        dstSz = pBufferCasted->epptr() - dstBegin;
+                        assert(dstSz);
+                    }
+
+                    read = 0;
+                    while (writerFunc(dstBegin, dstSz, read) && read > 0)
```

Comment:
instead of having two while loops they seem to have similar bodies, seems like we could use a do while here to avoid dry

Reply:
there are 1 while loop. While it is a bit ugly to have similar conditions, it looks to have a bit more clear logic.